### ISSUE TITLE
NXptycho format and CXI compatibility 

### ISSUE THREAD

**Issue by vincefn** [nexusformat/definitions#647]
Hi,
@aaron-parsons @FilipeMaia @vasole 
   I was glad to see that support for ptycho data has been contributed, but I have a few questions regarding the format:

1. is there a particular reason almost all fields are required ? e.g. beam divergence, etc... Compared to the CXI ptycho example which is absolutely minimal it may be risky to enforce having all fields required even though it's not necessary for interpretation

2. regarding the experiment energy. The CXI example uses source_1/energy to describe the "Energy of each photon", whereas for NXsource the energy is that of the source (a few GeV for synchrotrons). Any proposed way around this incompatibility (which is not just for ptycho but would apply to other CXI datasets).

3. The detector data in data_1 must have the shape (npts_x*npts_y, frame_size_x, frame_size_y). Since NeXus uses row-major (C) formatting, should that not be (npts_x*npts_y, frame_size_y, frame_size_x) ? I see this is the same in other definitions (sas, tomorrow) but I don't understand the rationale. Of course 'x' could be rows and 'y' columns but that feels strange.

4. What does NXsample/transformations@vector mean ?

**Comment by vasole** [comment]
>   What does NXsample/transformations@vector mean ?

It is the vector (of unity module) defining the translation direction or the rotation axis depending on the type of transformation

**Comment by vasole** [comment]
Concerning point 2, the energy in terms of CXI should be mapped to something of like incident_energy inside an NXbeam or energy inside an NXmonocromator.

**Comment by vasole** [comment]
@vincefn 

Looking at http://download.nexusformat.org/doc/html/classes/contributed_definitions/NXcxi_ptycho.html?highlight=nxptycho

It seems the ambiguity you are mentioning concerning the energy is solved in the current version of the page.

**Comment by vincefn** [comment]
> > What does NXsample/transformations@vector mean ?
> 
> It is the vector (of unity module) defining the translation direction or the rotation axis depending on the type of transformation

I understand the NXtransformation@vector definition, but not what this represents in NXptycho

**Comment by vincefn** [comment]
> @vincefn
> 
> Looking at http://download.nexusformat.org/doc/html/classes/contributed_definitions/NXcxi_ptycho.html?highlight=nxptycho
> 
> It seems the ambiguity you are mentioning concerning the energy is solved in the current version of the page.

The ambiguity does not lie in the NXptycho definition - which is clear. However the issue is with the CXI compatibility. But I think there is no way out since a 'source' is used in CXI for the energy.
In my implementation (only for CDI so far), I have added both a source_1:NXsource and a beam_1:NXbeam, listing respectively energy and incident_energy as the photon energy (which is wrong for the source but maintains CXI compatiblity)

**Comment by vasole** [comment]
Well, you can always find a way to make them compatible. Nothing prevents you in NeXus from naming source to something that is an NXbeam and not an NXsource. That would iply the definition might need to be updated.

As far as I understand cxi is name based while NeXus is attribute based. I can understand is something weird, but it can be adapted.

**Comment by vincefn** [comment]
> Well, you can always find a way to make them compatible. Nothing prevents you in NeXus from naming source to something that is an NXbeam and not an NXsource. That would iply the definition might need to be updated.

Except a NXbeam does not have an 'energy' field (contrary to NXsource), but only 'incident_energy' and 'final_energy'.

**Comment by vasole** [comment]
Just a reminder, the (non written?) rule is that it is allowed to add things to a NeXus group. It is not allowed to change what a dataset represents.

**Comment by benajamin** [comment]
>1. is there a particular reason almost all fields are required ? e.g. beam divergence, etc... Compared to the CXI ptycho example which is absolutely minimal it may be risky to enforce having all fields required even though it's not necessary for interpretation

I agree it looks overly restrictive. Could you please propose which specific fields should be made optional?

**Comment by aaron-parsons** [comment]
Hi folks, sorry for not replying sooner. You are correct, we can set some things to optional. This went through in a bit of rush so hence minOccurs and maxOccurs are set to 1. I'll take a look over the next week or so and get back to you.

Aaron

**Comment by vincefn** [comment]
@aaron-parsons I see you will also be at the ptycho-developer workshop in Berkeley early June, so this would be the ideal place to wrap this up.

**Comment by aaron-parsons** [comment]
Yes. The work was actually kicked off in the last one of these meetings, and the Berkeley folk were involved in the initial discussions. So definitely a good place to finish it off.

**Comment by aaron-parsons** [comment]
Hi @vincefn, I'm just looking through this ahead of our discussions over the next few days. Am I right in thinking that it's the NXbeam we want to open up a little. So make extent and incident_beam_divergence optional? Also, the NXmonitor and the beam centers in the NXdetector.

I think everything else is pretty standard, energy spread is useful and usually known.

Let me know if there is something else you want me to consider.

**Comment by vasole** [comment]
"""
An example here is that CXI expects the data to always to have shape (npts_x*npts_y, frame_size_x, frame_size_y). For nexus this is only true for arbitrary scan paths with raster format scans taking shape (npts_x, npts_y, frame_size_x, frame_size_y).
"""

Certainly I prefer the generic way. Well done for CXI.

Concerning the dimensions, is that truly what CXI says? I have always advocated to avoid using x and y and talk about dimension_0 and dimension1 to avoid ambiguities. To accommodate what is written one would have to transpose every frame in most of the cases (dimension_0=rows are mapped to y and dimension_1=columns to x). Is CXI advocating the use of the Fortran order?

For old stuff the damage is already done. For new stuff, I think we should take the opportunity to avoid ambiguities.

**Comment by vasole** [comment]
If a user community is providing an HDF5 based format, I see little interest in translating that format to NeXus. All what NeXus should do is to provide a container for it (typically an NXsubentry) in order

- to be able to refer to the original community supplied documentation and not a translation
- to allow software developers to easily support the format/definition irrespective of following the original or the NeXusfied version.

**Comment by FilipeMaia** [comment]
I believe the CXI format says `(npts, y, x)`, where the 'x' corresponds to the last (fastest changing) dimension of the data and, if left unspecified, is assumed to correspond be parallel to the x axis of the experiment.

**Comment by vasole** [comment]
> I believe the CXI format says (npts, y, x), where the 'x' corresponds to the last (fastest changing) dimension of the data and, if left unspecified, is assumed to correspond be parallel to the x axis of the experiment.

If that is the case then there is already a bug in the NeXus documentation and an example of the risk of getting lost in translation that can be avoided by taking the community supplied definition instead of trying to convert it.
